clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -454,64 +467,116 @@ private int append(Map<TopicPartition, List<ConsumerRecord<K, V>>> drained,
+                return RequestFuture.leaderNotAvailable();
+            }
+        }
+
+        // Group the partitions by node.
+        final Map<Node, Map<TopicPartition, Long>> timestampsToSearchByNode = new HashMap<>();
+        for (Map.Entry<TopicPartition, Long> entry: timestampsToSearch.entrySet()) {
+            Node node = metadata.fetch().leaderFor(entry.getKey());
+            Map<TopicPartition, Long> topicData = timestampsToSearchByNode.get(node);
+            if (topicData == null) {
+                topicData = new HashMap<>();
+                timestampsToSearchByNode.put(node, topicData);
+            }
+            topicData.put(entry.getKey(), entry.getValue());
+        }
+
+        final RequestFuture<Map<TopicPartition, OffsetAndTimestamp>> listOffsetRequestsFuture = new RequestFuture<>();
+        final Map<TopicPartition, OffsetAndTimestamp> timestampOffsets = new HashMap<>();
+        for (Map.Entry<Node, Map<TopicPartition, Long>> entry : timestampsToSearchByNode.entrySet()) {
+            sendListOffsetRequest(entry.getKey(), entry.getValue())
+                    .addListener(new RequestFutureListener<Map<TopicPartition, OffsetAndTimestamp>>() {
+                        @Override
+                        public void onSuccess(Map<TopicPartition, OffsetAndTimestamp> value) {
+                            timestampOffsets.putAll(value);
+                            if (timestampOffsets.size() == timestampsToSearch.size() && !listOffsetRequestsFuture.isDone())
+                                listOffsetRequestsFuture.complete(timestampOffsets);
+                        }
+
                         @Override
-                        public void onSuccess(ClientResponse response, RequestFuture<Long> future) {
-                            handleListOffsetResponse(topicPartition, response, future);
+                        public void onFailure(RuntimeException e) {
+                            // This may cause all the requests to be retried, but should be rare.
+                            if (!listOffsetRequestsFuture.isDone())
+                                listOffsetRequestsFuture.raise(e);
                         }
                     });
         }
+        return listOffsetRequestsFuture;
+    }
+
+    /**
+     * Send the ListOffsetRequest to a specific broker for the partitions and target timestamps.
+     *
+     * @param node The node of the
+     * @param timestampsToSearch The mapping from partitions to the targe timestamps.
+     * @return A response which can be polled to obtain the corresponding timestamps and offsets.
+     */
+    private RequestFuture<Map<TopicPartition, OffsetAndTimestamp>> sendListOffsetRequest(Node node,
+                                                                                      final Map<TopicPartition, Long> timestampsToSearch) {
+        ListOffsetRequest request = new ListOffsetRequest(timestampsToSearch, 1);
```

Comment:
We actually have a precedent for using static factory methods. The `FetchRequest` class  has a static method `fromReplica`. I guess I'm wondering why we're fighting so hard to use the constructors when there's already an easy way out. It seems like a step backwards to lose the (albeit marginal) safety provided by the generic types.